File: Punto3/Matriz.java
import java.util.Arrays;

public class Matriz {
    public final int filas;
    public final int columnas;
    private final double[] datos;
    public String tipoElemento; 

    public Matriz(int filas, int columnas, String tipoElemento) {
        this.filas = filas;
        this.columnas = columnas;
        this.tipoElemento = tipoElemento;
        this.datos = new double[filas * columnas];
    }

    public Matriz(int filas, int columnas, String tipoElemento, double[] datos) {
        if (datos.length != filas * columnas) {
            throw new IllegalArgumentException("Longitud de datos no coincide con dimensiones");
        }
        this.filas = filas;
        this.columnas = columnas;
        this.tipoElemento = tipoElemento;
        this.datos = Arrays.copyOf(datos, datos.length);
    }

    public double get(int i, int j) {
        return datos[i * columnas + j];
    }

    public void set(int i, int j, double val) {
        datos[i * columnas + j] = val;
    }

    public double[] datos() {
        return datos;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("[");
        for (int i = 0; i < filas; i++) {
            sb.append("[");
            for (int j = 0; j < columnas; j++) {
                double v = get(i, j);
                if ("entero".equals(tipoElemento)) {
                    sb.append((int) v);
                } else {
                    sb.append(v);
                }
                if (j < columnas - 1) sb.append(", ");
            }
            sb.append("]");
            if (i < filas - 1) sb.append(", ");
        }
        sb.append("]");
        return sb.toString();
    }

    public static Matriz crearCeros(int filas, int columnas, String tipoElemento) {
        return new Matriz(filas, columnas, tipoElemento);
    }

    public Matriz transpose() {
        Matriz t = new Matriz(columnas, filas, this.tipoElemento);
        for (int i = 0; i < filas; i++) {
            for (int j = 0; j < columnas; j++) {
                t.set(j, i, this.get(i, j));
            }
        }
        return t;
    }

    public static Matriz matmulTranspose(Matriz A, Matriz B) {
        if (A.columnas != B.filas) {
            throw new IllegalArgumentException(
                String.format("Dimensiones incompatibles para multiplicación: cols(A)=%d != filas(B)=%d", A.columnas, B.filas));
        }
        String tipoRes = promoverTipo(A.tipoElemento, B.tipoElemento);
        if (tipoRes == null) {
            throw new IllegalArgumentException("Tipos de elemento incompatibles: " + A.tipoElemento + " vs " + B.tipoElemento);
        }
        Matriz BT = B.transpose();
        int m = A.filas;
        int p = B.columnas;
        int n = A.columnas;
        Matriz C = new Matriz(m, p, tipoRes);
        for (int i = 0; i < m; i++) {
            int offA = i * A.columnas;
            for (int j = 0; j < p; j++) {
                int offBT = j * BT.columnas; // BT.columnas == B.filas == n
                double s = 0;
                for (int k = 0; k < n; k++) {
                    double a = A.datos()[offA + k];
                    double b = BT.datos()[offBT + k];
                    s += a * b;
                }
                if ("entero".equals(tipoRes)) {
                    C.datos()[i * p + j] = (int) Math.round(s);
                } else {
                    C.datos()[i * p + j] = s;
                }
            }
        }
        return C;
    }

    public static Matriz matmulBlock(Matriz A, Matriz B, int tile) {
        if (A.columnas != B.filas) {
            throw new IllegalArgumentException(
                String.format("Dimensiones incompatibles para multiplicación: cols(A)=%d != filas(B)=%d", A.columnas, B.filas));
        }
        String tipoRes = promoverTipo(A.tipoElemento, B.tipoElemento);
        if (tipoRes == null) {
            throw new IllegalArgumentException("Tipos de elemento incompatibles: " + A.tipoElemento + " vs " + B.tipoElemento);
        }
        int m = A.filas;
        int n = A.columnas;
        int p = B.columnas;
        Matriz C = new Matriz(m, p, tipoRes);
        for (int ii = 0; ii < m; ii += tile) {
            for (int jj = 0; jj < p; jj += tile) {
                for (int kk = 0; kk < n; kk += tile) {
                    int iLim = Math.min(ii + tile, m);
                    int jLim = Math.min(jj + tile, p);
                    int kLim = Math.min(kk + tile, n);
                    for (int i = ii; i < iLim; i++) {
                        for (int k = kk; k < kLim; k++) {
                            double a = A.get(i, k);
                            int baseC = i * p;
                            int baseB = k * p;
                            for (int j = jj; j < jLim; j++) {
                                C.datos()[baseC + j] += a * B.datos()[baseB + j];
                            }
                        }
                    }
                }
            }
        }
        if ("entero".equals(tipoRes)) {
            for (int i = 0; i < C.datos().length; i++) {
                C.datos()[i] = Math.round(C.datos()[i]);
            }
        }
        return C;
    }

    public static String promoverTipo(String t1, String t2) {
        if (t1.equals(t2)) return t1;
        if ((t1.equals("entero") && t2.equals("flotante")) || (t1.equals("flotante") && t2.equals("entero"))) return "flotante";
        return null;
    }
}

File: Punto2/matriz.py
from tipos import promover_tipo

class Matriz:
    def __init__(self, filas, columnas, tipo_elemento, datos=None):
        self.filas = filas
        self.columnas = columnas
        self.tipo_elemento = tipo_elemento
        if datos is None:
            self.datos = [0] * (filas * columnas)
        else:
            if len(datos) != filas * columnas:
                raise ValueError("Longitud de datos no coincide con dimensiones")
            self.datos = list(datos)

    def get(self, i, j):
        return self.datos[i * self.columnas + j]

    def set(self, i, j, val):
        self.datos[i * self.columnas + j] = val

    def copiar(self):
        return Matriz(self.filas, self.columnas, self.tipo_elemento, list(self.datos))

    def __repr__(self):
        filas = []
        for i in range(self.filas):
            fila = [str(self.get(i, j)) for j in range(self.columnas)]
            filas.append("[" + ", ".join(fila) + "]")
        return "[" + ", ".join(filas) + "]"

def crear_matriz_ceros(filas, columnas, tipo_elemento):
    return Matriz(filas, columnas, tipo_elemento)

def matriz_desde_literal(lista_de_filas):
    if not isinstance(lista_de_filas, list) or len(lista_de_filas) == 0:
        raise ValueError("Literal de matriz vacío o inválido")
    filas = len(lista_de_filas)
    columnas = None
    tipo_elem = None
    flat = []
    for i, fila in enumerate(lista_de_filas):
        if not isinstance(fila, list):
            raise ValueError(f"Fila {i} de la matriz literal no es una lista")
        if columnas is None:
            columnas = len(fila)
            if columnas == 0:
                raise ValueError("Filas de la matriz no pueden ser vacías")
        else:
            if len(fila) != columnas:
                raise ValueError(f"Literal de matriz no rectangular: fila {i} longitud {len(fila)} != {columnas}")
        for v in fila:
            if isinstance(v, float):
                t = 'flotante'
            elif isinstance(v, int):
                t = 'entero'
            else:
                raise ValueError("Valores de matriz deben ser numéricos (int o float)")
            if tipo_elem is None:
                tipo_elem = t
            else:
                tipo_elem = promover_tipo(tipo_elem, t) or tipo_elem
            flat.append(v)
    if tipo_elem is None:
        raise ValueError("No se pudo inferir tipo de matriz")
    if tipo_elem == 'flotante':
        flat = [float(x) for x in flat]
    return Matriz(filas, columnas, tipo_elem, flat)

def transpose(m):
    datos = [0] * (m.filas * m.columnas)
    for i in range(m.filas):
        for j in range(m.columnas):
            datos[j * m.filas + i] = m.get(i, j)
    return Matriz(m.columnas, m.filas, m.tipo_elemento, datos)

def matmul_transpose(A, B):
    if A.columnas != B.filas:
        raise ValueError(f"Dimensiones incompatibles para multiplicación: cols(A)={A.columnas} != filas(B)={B.filas}")
    tipo_res = promover_tipo(A.tipo_elemento, B.tipo_elemento)
    if tipo_res is None:
        raise ValueError(f"Tipos de elemento incompatibles: {A.tipo_elemento} vs {B.tipo_elemento}")
    B_T = transpose(B)
    m = A.filas
    p = B.columnas
    n = A.columnas
    C = crear_matriz_ceros(m, p, tipo_res)
    for i in range(m):
        offA = i * A.columnas
        for j in range(p):
            offBT = j * B_T.columnas  
            s = 0
            for k in range(n):
                a = A.datos[offA + k]
                b = B_T.datos[offBT + k]
                s += a * b
            C.datos[i * p + j] = float(s) if tipo_res == 'flotante' else int(s)
    return C

def matmul_block(A, B, tile=32):
    if A.columnas != B.filas:
        raise ValueError(f"Dimensiones incompatibles para multiplicación: cols(A)={A.columnas} != filas(B)={B.filas}")
    tipo_res = promover_tipo(A.tipo_elemento, B.tipo_elemento)
    if tipo_res is None:
        raise ValueError(f"Tipos de elemento incompatibles: {A.tipo_elemento} vs {B.tipo_elemento}")
    m, n, p = A.filas, A.columnas, B.columnas
    C = crear_matriz_ceros(m, p, tipo_res)
    for ii in range(0, m, tile):
        for jj in range(0, p, tile):
            for kk in range(0, n, tile):
                i_lim = min(ii + tile, m)
                j_lim = min(jj + tile, p)
                k_lim = min(kk + tile, n)
                for i in range(ii, i_lim):
                    for k in range(kk, k_lim):
                        a = A.get(i, k)
                        baseC = i * p
                        baseB = k * p
                        for j in range(jj, j_lim):
                            C.datos[baseC + j] += a * B.datos[baseB + j]
    if tipo_res == 'entero':
        C.datos = [int(x) for x in C.datos]
    else:
        C.datos = [float(x) for x in C.datos]
    return C